src/internal/proc.js

```diff
@@ -420,6 +408,28 @@ export default function proc(
     cb.cancel = takeCb.cancel
   }
 
+  function runDeclarativeEffect(effect, effectId, cb) {
+    if (options.middleware) {
```

Comment:
this one could have been checked before entering `runDeclarativeEffect` and `runDeclarativeEffect` should run only if this returns something other than `effect` itself